File: app/flyfun-forms/flyfun-forms/Services/MRZParser.swift
import Foundation
import RZUtilsSwift

struct MRZScanResult {
    let surname: String
    let givenNames: String
    let passportNumber: String
    let nationality: String
    let dateOfBirth: Date
    let expiryDate: Date
    let gender: String
    let issuingCountry: String
    let format: MRZFormat
}

enum MRZFormat { case td3, td1 }

enum MRZParser {

    // MARK: - Public API

    static func parse(lines: [String]) -> MRZScanResult? {
        let trimmed = lines.map { $0.trimmingCharacters(in: .whitespaces) }
        if trimmed.count == 2, trimmed[0].count == 44, trimmed[1].count == 44 {
            return parseTD3(line1: trimmed[0], line2: trimmed[1])
        }
        if trimmed.count == 3, trimmed[0].count == 30, trimmed[1].count == 30, trimmed[2].count == 30 {
            return parseTD1(line1: trimmed[0], line2: trimmed[1], line3: trimmed[2])
        }
        return nil
    }

    // MARK: - ICAO 9303 Check Digit

    static func checkDigit(_ input: String) -> Int {
        let weights = [7, 3, 1]
        var sum = 0
        for (i, ch) in input.enumerated() {
            let value: Int
            if ch == "<" {
                value = 0
            } else if ch.isNumber {
                value = Int(String(ch))!
            } else if ch.isUppercase {
                value = Int(ch.asciiValue! - Character("A").asciiValue!) + 10
            } else {
                value = 0
            }
            sum += value * weights[i % 3]
        }
        return sum % 10
    }

    // MARK: - OCR Sanitization (digits only)

    static func sanitizeDigits(_ s: String) -> String {
        var result = ""
        for ch in s {
            switch ch {
            case "O": result.append("0")
            case "I": result.append("1")
            case "B": result.append("8")
            case "S": result.append("5")
            case "G": result.append("6")
            case "Q": result.append("0")
            default: result.append(ch)
            }
        }
        return result
    }

    // MARK: - OCR Correction for Alphanumeric Fields

    /// When checksum fails on a field that may contain letters (e.g. passport number),
    /// try common digit→letter substitutions to find the correct reading.
    /// MRZ font makes D↔0, I↔1, B↔8, S↔5, G↔6, Z↔2 ambiguous.
    static func correctField(_ field: String, expectedCheck: Int) -> String? {
        // digit → possible letters (reverse of sanitizeDigits + common MRZ confusions)
        let substitutions: [Character: [Character]] = [
            "0": ["O", "D", "Q"],
            "1": ["I", "L"],
            "8": ["B"],
            "5": ["S"],
            "6": ["G"],
            "2": ["Z"],
        ]

        // Try single-character corrections first (most common case)
        var chars = Array(field)
        for i in 0..<chars.count {
            let original = chars[i]
            if let alternatives = substitutions[original] {
                for alt in alternatives {
                    chars[i] = alt
                    let candidate = String(chars)
                    if checkDigit(candidate) == expectedCheck {
                        return candidate
                    }
                }
                chars[i] = original
            }
        }

        // Try two-character corrections (for cases like D+I both misread)
        for i in 0..<chars.count {
            let origI = chars[i]
            guard let altsI = substitutions[origI] else { continue }
            for altI in altsI {
                chars[i] = altI
                for j in (i+1)..<chars.count {
                    let origJ = chars[j]
                    guard let altsJ = substitutions[origJ] else { continue }
                    for altJ in altsJ {
                        chars[j] = altJ
                        let candidate = String(chars)
                        if checkDigit(candidate) == expectedCheck {
                            return candidate
                        }
                    }
                    chars[j] = origJ
                }
                chars[i] = origI
            }
        }

        return nil
    }

    // MARK: - Date Parsing

    static func parseMRZDate(_ s: String, isBirthDate: Bool) -> Date? {
        guard s.count == 6 else { return nil }
        guard let yy = Int(String(s.prefix(2))),
              let mm = Int(String(s.dropFirst(2).prefix(2))),
              let dd = Int(String(s.suffix(2))) else { return nil }
        guard mm >= 1, mm <= 12, dd >= 1, dd <= 31 else { return nil }

        let century: Int
        if isBirthDate {
            let currentYY = Calendar.current.component(.year, from: Date()) % 100
            century = yy > currentYY ? 1900 : 2000
        } else {
            century = 2000
        }

        var comps = DateComponents()
        comps.year = century + yy
        comps.month = mm
        comps.day = dd
        comps.calendar = Calendar(identifier: .gregorian)
        return comps.date
    }

    // MARK: - TD3 (Passport, 2×44)

    private static func parseTD3(line1: String, line2: String) -> MRZScanResult? {
        // Sanitize only digit-expected positions: check digits (9,19,27,43) and dates (13-18, 21-26)
        let l2 = sanitizePositions(line2, positions: [9, 13, 14, 15, 16, 17, 18, 19, 21, 22, 23, 24, 25, 26, 27, 43])

        // Validate check digits
        var passNum = substr(l2, 0, 9)
        let passCheck = checkDigit(passNum)
        let passExpected = digit(l2, 9)
        if passCheck != passExpected {
            // Try OCR correction (digit→letter substitutions)
            if let corrected = correctField(passNum, expectedCheck: passExpected) {
                RZSLog.info("MRZ passport corrected: \(passNum) → \(corrected)")
                passNum = corrected
            } else {
                RZSLog.info("MRZ passport check FAIL: \(passNum) computed=\(passCheck) expected=\(passExpected)")
                return nil
            }
        }

        let dob = substr(l2, 13, 6)
        let dobCheck = checkDigit(dob)
        let dobExpected = digit(l2, 19)
        guard dobCheck == dobExpected else {
            RZSLog.info("MRZ DOB check FAIL: \(dob) computed=\(dobCheck) expected=\(dobExpected)")
            return nil
        }

        let expiry = substr(l2, 21, 6)
        let expiryCheck = checkDigit(expiry)
        let expiryExpected = digit(l2, 27)
        guard expiryCheck == expiryExpected else {
            RZSLog.info("MRZ expiry check FAIL: \(expiry) computed=\(expiryCheck) expected=\(expiryExpected)")
            return nil
        }

        // Composite check: positions 0-9 + 13-19 + 21-27 + 28-42
        // Use corrected passport number (passNum) + its check digit for positions 0-9
        let compositeInput = passNum + String(substr(l2, 9, 1)) + substr(l2, 13, 7) + substr(l2, 21, 7) + substr(l2, 28, 15)
        let compCheck = checkDigit(compositeInput)
        let compExpected = digit(l2, 43)
        guard compCheck == compExpected else {
            RZSLog.info("MRZ composite check FAIL: computed=\(compCheck) expected=\(compExpected)")
            return nil
        }

        // Parse names from line 1
        let (surname, givenNames) = parseNames(String(line1.dropFirst(5)))
        let issuingCountry = cleanCountry(substr(line1, 2, 3))
        let nationality = cleanCountry(substr(l2, 10, 3))
        let gender = parseGender(substr(l2, 20, 1))

        guard let birthDate = parseMRZDate(dob, isBirthDate: true),
              let expiryDate = parseMRZDate(expiry, isBirthDate: false) else { return nil }

        return MRZScanResult(
            surname: surname,
            givenNames: givenNames,
            passportNumber: cleanField(passNum),
            nationality: nationality,
            dateOfBirth: birthDate,
            expiryDate: expiryDate,
            gender: gender,
            issuingCountry: issuingCountry,
            format: .td3
        )
    }

    // MARK: - TD1 (ID Card, 3×30)

    private static func parseTD1(line1: String, line2: String, line3: String) -> MRZScanResult? {
        // Sanitize only digit-expected positions
        // Line 1: check digit at 14
        let l1 = sanitizePositions(line1, positions: [14])
        // Line 2: DOB 0-5, check 6, expiry 8-13, check 14, composite check 29
        let l2 = sanitizePositions(line2, positions: [0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 13, 14, 29])

        let docNum = substr(l1, 5, 9)
        guard checkDigit(docNum) == digit(l1, 14) else { return nil }

        let dob = substr(l2, 0, 6)
        guard checkDigit(dob) == digit(l2, 6) else { return nil }

        let expiry = substr(l2, 8, 6)
        guard checkDigit(expiry) == digit(l2, 14) else { return nil }

        // Composite: line1[5..29] + line2[0..6] + line2[8..14] + line2[18..28]
        let compositeInput = substr(l1, 5, 25) + substr(l2, 0, 7) + substr(l2, 8, 7) + substr(l2, 18, 11)
        guard checkDigit(compositeInput) == digit(l2, 29) else { return nil }

        let issuingCountry = cleanCountry(substr(l1, 2, 3))
        let nationality = cleanCountry(substr(l2, 15, 3))
        let gender = parseGender(substr(l2, 7, 1))

        let (surname, givenNames) = parseNames(line3)

        guard let birthDate = parseMRZDate(dob, isBirthDate: true),
              let expiryDate = parseMRZDate(expiry, isBirthDate: false) else { return nil }

        return MRZScanResult(
            surname: surname,
            givenNames: givenNames,
            passportNumber: cleanField(docNum),
            nationality: nationality,
            dateOfBirth: birthDate,
            expiryDate: expiryDate,
            gender: gender,
            issuingCountry: issuingCountry,
            format: .td1
        )
    }

    // MARK: - Helpers

    private static func substr(_ s: String, _ start: Int, _ length: Int) -> String {
        let startIdx = s.index(s.startIndex, offsetBy: start)
        let endIdx = s.index(startIdx, offsetBy: length)
        return String(s[startIdx..<endIdx])
    }

    private static func digit(_ s: String, _ pos: Int) -> Int {
        let ch = s[s.index(s.startIndex, offsetBy: pos)]
        return Int(String(ch)) ?? -1
    }

    /// Apply OCR digit sanitization only to specific positions (dates, check digits).
    private static func sanitizePositions(_ line: String, positions: Set<Int>) -> String {
        var chars = Array(line)
        for i in positions where i < chars.count {
            let sanitized = sanitizeDigits(String(chars[i]))
            chars[i] = sanitized.first!
        }
        return String(chars)
    }

    private static func parseNames(_ nameField: String) -> (surname: String, givenNames: String) {
        let parts = nameField.components(separatedBy: "<<")
        let surname = titleCase(parts[0].replacingOccurrences(of: "<", with: " ").trimmingCharacters(in: .whitespaces))
        let given: String
        if parts.count > 1 {
            given = titleCase(parts[1...].joined(separator: " ").replacingOccurrences(of: "<", with: " ").trimmingCharacters(in: .whitespaces))
        } else {
            given = ""
        }
        return (surname, given)
    }

    private static func titleCase(_ s: String) -> String {
        s.lowercased().split(separator: " ").map { $0.prefix(1).uppercased() + $0.dropFirst() }.joined(separator: " ")
    }

    private static func cleanField(_ s: String) -> String {
        s.replacingOccurrences(of: "<", with: "").trimmingCharacters(in: .whitespaces)
    }

    private static func cleanCountry(_ s: String) -> String {
        s.replacingOccurrences(of: "<", with: "").trimmingCharacters(in: .whitespaces)
    }

    private static func parseGender(_ s: String) -> String {
        switch s {
        case "M": return "M"
        case "F": return "F"
        default: return "X"
        }
    }
}

File: app/android/core-logic/src/main/kotlin/aero/flyfun/forms/logic/MRZParser.kt
package aero.flyfun.forms.logic

import java.time.DateTimeException
import java.time.LocalDate

enum class MRZFormat { TD3, TD1 }

data class MRZScanResult(
    val surname: String,
    val givenNames: String,
    val passportNumber: String,
    val nationality: String,
    val dateOfBirth: LocalDate,
    val expiryDate: LocalDate,
    val gender: String,
    val issuingCountry: String,
    val format: MRZFormat,
)

/**
 * ICAO 9303 machine-readable-zone parser.
 *
 * Port of `Services/MRZParser.swift`; the Swift tests are the specification.
 * Pure logic - OCR (Vision on iOS, ML Kit on Android) supplies the lines.
 */
object MRZParser {

    fun parse(lines: List<String>, today: LocalDate = LocalDate.now()): MRZScanResult? {
        val trimmed = lines.map { it.trim() }
        if (trimmed.size == 2 && trimmed[0].length == 44 && trimmed[1].length == 44) {
            return parseTD3(trimmed[0], trimmed[1], today)
        }
        if (trimmed.size == 3 && trimmed.all { it.length == 30 }) {
            return parseTD1(trimmed[0], trimmed[1], trimmed[2], today)
        }
        return null
    }

    // MARK: ICAO 9303 check digit

    fun checkDigit(input: String): Int {
        val weights = intArrayOf(7, 3, 1)
        var sum = 0
        input.forEachIndexed { i, ch ->
            val value = when {
                ch == '<' -> 0
                ch.isDigit() -> ch - '0'
                ch in 'A'..'Z' -> ch - 'A' + 10
                else -> 0
            }
            sum += value * weights[i % 3]
        }
        return sum % 10
    }

    // MARK: OCR sanitization (digit-expected positions only)

    fun sanitizeDigits(s: String): String = buildString {
        for (ch in s) {
            append(
                when (ch) {
                    'O' -> '0'
                    'I' -> '1'
                    'B' -> '8'
                    'S' -> '5'
                    'G' -> '6'
                    'Q' -> '0'
                    else -> ch
                }
            )
        }
    }

    /**
     * When the checksum fails on a field that may contain letters (a passport
     * number, say), try the substitutions the MRZ font makes ambiguous:
     * D/0, I/1, B/8, S/5, G/6, Z/2.
     */
    fun correctField(field: String, expectedCheck: Int): String? {
        val substitutions: Map<Char, List<Char>> = mapOf(
            '0' to listOf('O', 'D', 'Q'),
            '1' to listOf('I', 'L'),
            '8' to listOf('B'),
            '5' to listOf('S'),
            '6' to listOf('G'),
            '2' to listOf('Z'),
        )

        val chars = field.toCharArray()

        // Single-character corrections first - much the most common case.
        for (i in chars.indices) {
            val original = chars[i]
            val alternatives = substitutions[original] ?: continue
            for (alt in alternatives) {
                chars[i] = alt
                val candidate = String(chars)
                if (checkDigit(candidate) == expectedCheck) return candidate
            }
            chars[i] = original
        }

        // Then pairs, for cases where two characters were both misread.
        for (i in chars.indices) {
            val origI = chars[i]
            val altsI = substitutions[origI] ?: continue
            for (altI in altsI) {
                chars[i] = altI
                for (j in (i + 1) until chars.size) {
                    val origJ = chars[j]
                    val altsJ = substitutions[origJ] ?: continue
                    for (altJ in altsJ) {
                        chars[j] = altJ
                        val candidate = String(chars)
                        if (checkDigit(candidate) == expectedCheck) return candidate
                    }
                    chars[j] = origJ
                }
                chars[i] = origI
            }
        }

        return null
    }

    // MARK: Dates

    /**
     * MRZ dates are YYMMDD with no century. A birth date whose two-digit year is
     * greater than the current one must be last century; an expiry is always this
     * one.
     *
     * [today] is injectable so the century rule can be tested without the result
     * depending on when the suite runs.
     */
    fun parseMRZDate(s: String, isBirthDate: Boolean, today: LocalDate = LocalDate.now()): LocalDate? {
        if (s.length != 6) return null
        val yy = s.substring(0, 2).toIntOrNull() ?: return null
        val mm = s.substring(2, 4).toIntOrNull() ?: return null
        val dd = s.substring(4, 6).toIntOrNull() ?: return null
        if (mm !in 1..12 || dd !in 1..31) return null

        val century = if (isBirthDate) {
            if (yy > today.year % 100) 1900 else 2000
        } else {
            2000
        }

        // Swift's Calendar is lenient and rolls 31 February into March; LocalDate
        // rejects it. Rejecting is the better answer for a scanned document, and
        // no ported test depends on the rollover.
        return try {
            LocalDate.of(century + yy, mm, dd)
        } catch (_: DateTimeException) {
            null
        }
    }

    // MARK: TD3 (passport, 2x44)

    private fun parseTD3(line1: String, line2: String, today: LocalDate): MRZScanResult? {
        // Sanitize only the positions that must be digits: check digits and dates.
        val l2 = sanitizePositions(
            line2,
            setOf(9, 13, 14, 15, 16, 17, 18, 19, 21, 22, 23, 24, 25, 26, 27, 43),
        )

        var passNum = substr(l2, 0, 9)
        val passExpected = digit(l2, 9)
        if (checkDigit(passNum) != passExpected) {
            passNum = correctField(passNum, passExpected) ?: return null
        }

        val dob = substr(l2, 13, 6)
        if (checkDigit(dob) != digit(l2, 19)) return null

        val expiry = substr(l2, 21, 6)
        if (checkDigit(expiry) != digit(l2, 27)) return null

        // Composite check over 0-9 + 13-19 + 21-27 + 28-42, using the corrected
        // passport number.
        val compositeInput = passNum + substr(l2, 9, 1) + substr(l2, 13, 7) +
            substr(l2, 21, 7) + substr(l2, 28, 15)
        val trailingTruncated = substr(l2, 28, 16).all { it == '<' }
        if (!trailingTruncated && checkDigit(compositeInput) != digit(l2, 43)) return null

        val (surname, givenNames) = parseNames(line1.drop(5))
        val birthDate = parseMRZDate(dob, isBirthDate = true, today = today) ?: return null
        val expiryDate = parseMRZDate(expiry, isBirthDate = false, today = today) ?: return null

        return MRZScanResult(
            surname = surname,
            givenNames = givenNames,
            passportNumber = cleanField(passNum),
            nationality = cleanCountry(substr(l2, 10, 3)),
            dateOfBirth = birthDate,
            expiryDate = expiryDate,
            gender = parseGender(substr(l2, 20, 1)),
            issuingCountry = cleanCountry(substr(line1, 2, 3)),
            format = MRZFormat.TD3,
        )
    }

    // MARK: TD1 (ID card, 3x30)

    private fun parseTD1(line1: String, line2: String, line3: String, today: LocalDate): MRZScanResult? {
        val l1 = sanitizePositions(line1, setOf(14))
        val l2 = sanitizePositions(line2, setOf(0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 13, 14, 29))

        val docNum = substr(l1, 5, 9)
        if (checkDigit(docNum) != digit(l1, 14)) return null

        val dob = substr(l2, 0, 6)
        if (checkDigit(dob) != digit(l2, 6)) return null

        val expiry = substr(l2, 8, 6)
        if (checkDigit(expiry) != digit(l2, 14)) return null

        val compositeInput = substr(l1, 5, 25) + substr(l2, 0, 7) + substr(l2, 8, 7) + substr(l2, 18, 11)
        if (checkDigit(compositeInput) != digit(l2, 29)) return null

        val (surname, givenNames) = parseNames(line3)
        val birthDate = parseMRZDate(dob, isBirthDate = true, today = today) ?: return null
        val expiryDate = parseMRZDate(expiry, isBirthDate = false, today = today) ?: return null

        return MRZScanResult(
            surname = surname,
            givenNames = givenNames,
            passportNumber = cleanField(docNum),
            nationality = cleanCountry(substr(l2, 15, 3)),
            dateOfBirth = birthDate,
            expiryDate = expiryDate,
            gender = parseGender(substr(l2, 7, 1)),
            issuingCountry = cleanCountry(substr(l1, 2, 3)),
            format = MRZFormat.TD1,
        )
    }

    // MARK: Helpers

    private fun substr(s: String, start: Int, length: Int): String {
        if (start >= s.length) return ""
        return s.substring(start, minOf(start + length, s.length))
    }

    private fun digit(s: String, pos: Int): Int {
        if (pos >= s.length) return -1
        return s[pos].digitToIntOrNull() ?: -1
    }

    /** Apply OCR digit sanitization only to the given positions. */
    private fun sanitizePositions(line: String, positions: Set<Int>): String {
        val chars = line.toCharArray()
        for (i in positions) {
            if (i < chars.size) chars[i] = sanitizeDigits(chars[i].toString())[0]
        }
        return String(chars)
    }

    private fun parseNames(nameField: String): Pair<String, String> {
        val parts = nameField.split("<<")
        val surname = titleCase(parts[0].replace('<', ' ').trim())
        val given = if (parts.size > 1) {
            titleCase(parts.drop(1).joinToString(" ").replace('<', ' ').trim())
        } else {
            ""
        }
        return surname to given
    }

    private fun titleCase(s: String): String =
        s.lowercase()
            .split(" ")
            .filter { it.isNotEmpty() }
            .joinToString(" ") { it.replaceFirstChar { c -> c.uppercaseChar() } }

    private fun cleanField(s: String): String = s.replace("<", "").trim()

    private fun cleanCountry(s: String): String = s.replace("<", "").trim()

    private fun parseGender(s: String): String = when (s) {
        "M" -> "M"
        "F" -> "F"
        else -> "X"
    }
}
